File: 240415/행렬 곱셈 순서/matrix-multiplication-order.js
const input = require("fs").readFileSync(0).toString().split("\n")
const N = Number(input[0])
const nums = []
for (let i = 1; i <= N ; i++){
    if (i == 1){
        nums.push(Number(input[i].split(" ")[0]))
        nums.push(Number(input[i].split(" ")[1]))
    }
    else{
        nums.push(Number(input[i].split(" ")[1]))
    }
}

const dp = []

for (let i = 0; i <= N; i++){
    dp.push(Array(N+1).fill(0))
}


// for (let i = 0 ; i<= N; i++) dp[i][i] = nums[i]


for (let cnt = 0 ; cnt <= N ; cnt++){
    for (let i = 1; i <= N-cnt ; i++){
        const j = i + cnt
        if (cnt > 0){
            let minv = 500**3
            for (let p = i; p < j ; p++){
                minv = Math.min(minv, dp[i][p]+dp[p+1][j]+nums[i-1]*nums[p]*nums[j])
            }
            dp[i][j] = minv
        }
    }
}
// console.log(nums)
// console.log(dp)
console.log(dp[1][N])

File: 240415/행렬 곱셈 순서/matrix-multiplication-order.py
import sys
input = sys.stdin.readline


N = int(input())
nums = [0] * (N+1)

for i in range(N):
    a, b = map(int, input().split())
    nums[i] = a
    nums[i+1] = b

dp = [[0] * (N+1) for _ in range(N+1)]

for cnt in range(1, N+1):
    for i in range(1, N+1-cnt):
        j = i + cnt
        minv = 500**3
        for p in range(i, j):
            minv = min(minv, dp[i][p] + dp[p+1][j] + nums[i-1]* nums[p] * nums[j])
        
        dp[i][j] = minv

print(dp[1][N])